### PR TITLE
Fix support generic struct arrays with generic TypeSpecs

### PR DESCRIPTION
**Pull request by josesimoes**
## Description
- ResolveToken for a TypeDef instance now resolves generic TypeSpec tokens to their open generic TypeDef, enabling newarr, ldelem.any and stelem.any to succeed on generic struct element types.
- InitializeReference now properly handles a VAR field without generic instance context, falling back to Object instead of faiing, mirroring the existing MVAR behaviour.
- Narrow VAR fallback to array pre-allocation path only.
  + Add allowUnresolvedVarFallback parameter to InitializeReference.
  + Now using this when resolving VAR.
  + Update callers, noticably NewObject as this is the only path reached from ClearElements → NewObjectFromIndex during open-generic array pre-allocation.

## Motivation and Context
- newarr `Pair<TKey,TValue>` inside a generic method emits a generic TypeSpec token. The runtime had no handling for this token class in ResolveToken, causing CLR_E_WRONG_TYPE for newarr, ldelem.any and stelem.any on generic struct arrays.
Even after resolving the element type, pre-allocating struct slots in the new array calls InitializeReference on each field. For open generic structs the field types are unresolved VAR parameters; without a closed generic context at array-creation time the call unconditionally returned CLR_E_FAIL. Fields are written correctly by subsequent stfld instructions, so initialising them as objects (which are null) is safe.
- Edge case surfaced when adding new test code in MDP nanoframework/metadata-processor#232.
- Related with nanoframework/Home#782.

## How Has This Been Tested?
- [tested against nanoframework/CoreLibrary#245].

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [x] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dev Containers (changes related with Dev Containers, has no impact on code or features)
- [ ] Dependencies/declarations (update dependencies or assembly declarations and changes associated, has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist

- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).

### CODE REVIEW

**Comment by josesimoes** [comment]
@coderabbitai review

File: src/CLR/Core/Execution.cpp [hunk 1]
```diff
@@ -2117,13 +2117,18 @@ HRESULT CLR_RT_ExecutionEngine::InitializeReference(
...
+            {
+                NANOCLR_CHECK_HRESULT(
+                    ResolveGenericTypeParameter(*genericInstance, res.GenericParamPosition, realTypeDef, dt));
 
-            NANOCLR_CHECK_HRESULT(
-                ResolveGenericTypeParameter(*genericInstance, res.GenericParamPosition, realTypeDef, dt));
-
-            goto process_datatype;
+                goto process_datatype;
+            }
```
[the remark this answers is not in the document]
Comment (josesimoes) [reply src/CLR/Core/Execution.cpp:2140]: Fixed

[the remark this answers is not in the document]
Comment (josesimoes) [reply src/CLR/Core/Execution.cpp:2140]: Done
